agent/structs/config_entry_gateways.go

```diff
@@ -99,14 +100,29 @@ type GatewayTLSConfig struct {
 
 	// SDS allows configuring TLS certificate from an SDS service.
 	SDS *GatewayTLSSDSConfig `json:",omitempty"`
+
+	TLSMinVersion types.TLSVersion `json:",omitempty"`
+	TLSMaxVersion types.TLSVersion `json:",omitempty"`
+
+	// Define a subset of cipher suites to restrict
+	// Only applicable to connections negotiated via TLS 1.2 or earlier
+	CipherSuites []types.TLSCipherSuite `json:",omitempty"`
 }
 
+// TODO: is there a better way to define a common subset of fields shared by this struct and GatewayTLSConfig?
```

Comment:
Do you think this subset of fields will be used in other places? You can define the subset as a struct and embed it inside `GatewayServiceTLSConfig` and `GatewayTLSConfig` like
```go
type TLSVersionAndCipherConfig struct {
        TLSMinVersion types.TLSVersion `json:",omitempty"`
	TLSMaxVersion types.TLSVersion `json:",omitempty"`

	// Define a subset of cipher suites to restrict
	// Only applicable to connections negotiated via TLS 1.2 or earlier
	CipherSuites []types.TLSCipherSuite `json:",omitempty"`
}

type GatewayTLSConfig struct {
        TLSVersionAndCipherConfig // might have to double check how marshalling works with embed

	// Indicates that TLS should be enabled for this gateway or listener
	Enabled bool

	// SDS allows configuring TLS certificate from an SDS service.
	SDS *GatewayTLSSDSConfig `json:",omitempty"`
}

type GatewayServiceTLSConfig struct {
        TLSVersionAndCipherConfig // might have to double check how marshalling works with embed

        // Note no Enabled field here since it doesn't make sense to disable TLS on
	// one host on a TLS-configured listener.

	// SDS allows configuring TLS certificate from an SDS service.
	SDS *GatewayTLSSDSConfig `json:",omitempty"`
}
```

if it's only 2 instances of usage it might be better off being repeated until there's a reason to refactor it out 🤷

Reply:
Oh interesting, I hadn't seen that embedded struct pattern before! If there's potential complications with embedded JSON marshaling I'll leave it as-is for now, but will keep this in mind for the future!